File: data/snippets/quiz009.cpp
class Vehicle {
public:
    virtual void startEngine() {
        std::cout << "Engine started." << std::endl;
    }
    
    virtual void stopEngine() {
        std::cout << "Engine stopped." << std::endl;
    }
    
    virtual void accelerate() {
        std::cout << "Accelerating..." << std::endl;
    }
    
    virtual void brake() {
        std::cout << "Braking..." << std::endl;
    }
    
    virtual void refuel() {
        std::cout << "Refueling with gasoline." << std::endl;
    }
    
    virtual void changeGear() {
        std::cout << "Changing gear." << std::endl;
    }
};

class ElectricBicycle : public Vehicle {
public:
    // 电动自行车没有引擎，覆盖并拒绝父类方法
    void startEngine() override {
        throw std::logic_error("Electric bicycles don't have engines!");
    }
    
    void stopEngine() override {
        throw std::logic_error("Electric bicycles don't have engines!");
    }
    
    // 电动自行车不使用汽油
    void refuel() override {
        throw std::logic_error("Electric bicycles don't use gasoline!");
    }
    
    // 大多数电动自行车没有变速箱
    void changeGear() override {
        throw std::logic_error("This electric bicycle doesn't have gears!");
    }
    
    // 添加电动自行车特有的方法
    void chargeBattery() {
        std::cout << "Charging battery." << std::endl;
    }
    
    // 正常使用的方法
    void accelerate() override {
        std::cout << "Electric bicycle accelerating." << std::endl;
    }
    
    void brake() override {
        std::cout << "Electric bicycle braking." << std::endl;
    }
};

File: data/snippets/quiz009.java
class Vehicle {
    public void startEngine() {
        System.out.println("Engine started.");
    }
    
    public void stopEngine() {
        System.out.println("Engine stopped.");
    }
    
    public void accelerate() {
        System.out.println("Accelerating...");
    }
    
    public void brake() {
        System.out.println("Braking...");
    }
    
    public void refuel() {
        System.out.println("Refueling with gasoline.");
    }
    
    public void changeGear() {
        System.out.println("Changing gear.");
    }
    
    public void turnOnHeadlights() {
        System.out.println("Headlights turned on.");
    }
    
    public void honkHorn() {
        System.out.println("Horn honked.");
    }
}

class ElectricBicycle extends Vehicle {
    
    @Override
    public void startEngine() {
        throw new UnsupportedOperationException("Electric bicycles don't have engines!");
    }
    
    @Override
    public void stopEngine() {
        throw new UnsupportedOperationException("Electric bicycles don't have engines!");
    }
    
    @Override
    public void refuel() {
        throw new UnsupportedOperationException("Electric bicycles don't use gasoline!");
    }
    
    @Override
    public void changeGear() {
        throw new UnsupportedOperationException("This electric bicycle doesn't have gears!");
    }
    
    @Override
    public void honkHorn() {
        throw new UnsupportedOperationException("Electric bicycles typically don't have horns!");
    }
    
    public void chargeBattery() {
        System.out.println("Charging battery.");
    }
    
    public void checkBatteryLevel() {
        System.out.println("Battery level: 75%");
    }
    
    public void activateAssistMode() {
        System.out.println("Pedal assist mode activated.");
    }
    
    @Override
    public void accelerate() {
        System.out.println("Electric bicycle accelerating with motor assist.");
    }
    
    @Override
    public void brake() {
        System.out.println("Electric bicycle braking with regenerative charging.");
    }
    
    @Override
    public void turnOnHeadlights() {
        System.out.println("LED headlight turned on.");
    }
}
